[GTK][WPE] Add rendering and buffer formats information to webkit://gpu

carlosgcampos opened the pull request:
#### e36b742a17b74dd86287584684457d32e4dcf332
<pre>
[GTK][WPE] Add rendering and buffer formats information to webkit://gpu
<a href="https://bugs.webkit.org/show_bug.cgi?id=301623">https://bugs.webkit.org/show_bug.cgi?id=301623</a>

Reviewed by Adrian Perez de Castro.

Add information about threaded rendering, MSAA, preferred buffer formats
and supported buffer formats. This patch laso fixes the JSON output
that I broke in 302185@main.

Canonical link: <a href="https://commits.webkit.org/302293@main">https://commits.webkit.org/302293@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f3682267d4fe0142b7ca1fac962f0c0b8f3fea8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128644 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/910 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136033 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80047 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/857 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/785 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97931 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/65845 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/635 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115261 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78550 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/127995 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/571 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79314 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109017 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33850 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138486 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/738 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106467 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/785 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111600 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106290 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/622 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30130 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53105 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20091 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/798 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64044 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/664 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/721 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/747 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
```diff
@@ -212,68 +210,78 @@ static String dmabufRendererWithSupportedBuffers()
-    code[3] = (char) ((format >> 24) & 0xFF);
-    code[4] = '\0';
+    std::span<char> buffer;
+    CString code = CString::newUninitialized(4, buffer);
+    buffer[0] = static_cast<char>((format >> 0) & 0xFF);
+    buffer[1] = static_cast<char>((format >> 8) & 0xFF);
+    buffer[2] = static_cast<char>((format >> 16) & 0xFF);
+    buffer[3] = static_cast<char>((format >> 24) & 0xFF);
```
This feels like it could be:

```suggestion
    auto code = FourCC(format & 0xFFFFFFFF).string();
```

carlosgcampos replied:
It doesn't work, I see other names with this, I guess `format & 0xFFFFFFFF` is not what we should be passing to FourCC

carlosgcampos replied:
It's not critical, but with the CString approach we save an allocation, this returns a std::array and then I have to create a String from it to pass it to makeString, while makeString can handle CString directly

carlosgcampos committed:
[GTK][WPE] Add rendering and buffer formats information to webkit://gpu
https://bugs.webkit.org/show_bug.cgi?id=301623

Reviewed by Adrian Perez de Castro.

Add information about threaded rendering, MSAA, preferred buffer formats
and supported buffer formats. This patch laso fixes the JSON output
that I broke in 302185@main.

Canonical link: https://commits.webkit.org/302293@main
